subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/MultipleCandidateMatcher.java

```diff
@@ -246,72 +268,50 @@ private void removeCandidatesWithValueNotIn(int a, Set<Object> matchedValues) {
         return builder.build();
     }
 
-    private void setRequestedValue(int a, Object value) {
-        attributeValues[a] = value;
+    private Attribute<?> getAttribute(int a) {
+        if (a < requestedAttributes.length) {
+            return requestedAttributes[a];
+        } else {
+            Attribute<?> extraAttribute = extraAttributes[a - requestedAttributes.length];
+            Attribute<?> schemaAttribute = schema.getAttribute(extraAttribute.getName());
```

Comment:
It feels wrong that I have to do this here - Why don't we convert the attributes to their strongly typed counterparts when constructing the metadata? I think this part of the code shouldn't have to deal with untyped attributes coming from the repo unless they are completely unknown to the schema.

Reply:
I agree that this adding quite some complexity, not only here but also in other parts of the matching code. Which has bitten us several times already.
The problem, if I understand everything correctly, is that we require the _consumer_ schema to provide the typed version of the attribute, as is is the only one potentially knowing about the type. We do not know the _consumer_ when we construct the metadata based on a `.module` files. See: 
https://github.com/gradle/gradle/blob/5125ca11ee4f507f5db310f42b413d7858ba28a1/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/parser/ModuleMetadataParser.java#L323
The producer side does not know about the type in this case, as schemas or attribute types are not part of the published metadata.

Reply:
Let me see if I understand that correctly: Is it because we cache the metadata globally, so all clients of some module see the same instances?

Reply:
Yes.

I think the earliest point we _could_ coerce is when we transfer the (global) metadata into `ModuleComponentResolveMetadata` objects.